### ISSUE TITLE
`strstr()` a lot slower with ASAN (Linux glibc x86-64)

### ISSUE THREAD

**Issue by fuhsnn** [google/sanitizers#1832]
After enabling ASAN in the compiler project I work on, I noticed a `strstr()` call in the tokenization loop is responsible for significant slowdown in ASAN builds.

I believe this is worth reporting since it's a simple change that caused a huge diff, which should make an interesting case for profiling.

The worst case is about 170x slower on Skylake, 150x on Zen4, after replacing the `strstr()`  with a loop it's normal 2.8x / 2.5x. 
The change is https://github.com/fuhsnn/slimcc/commit/eb66a4619f168af7ce5c34ecc8cc37c82988d51d.

A script testing ASAN on-off and `strstr()` with-without, should work on any recent-ish glibc docker:
```
set -e

git clone https://github.com/fuhsnn/c_files
git clone https://github.com/fuhsnn/slimcc

cd slimcc
git checkout eb66a4619f168af7ce5c34ecc8cc37c82988d51d

cc *.c -o slimcc
echo "Asan 0, strstr 0"
time ./slimcc ../c_files/gcc.c -E -o/dev/null

cc *.c -o slimcc -fsanitize=address
echo "Asan 1 strstr 0"
time ./slimcc ../c_files/gcc.c -E -o/dev/null

git checkout HEAD~1

cc *.c -o slimcc
echo "Asan 0, strstr 1"
time ./slimcc ../c_files/gcc.c -E -o/dev/null

cc *.c -o slimcc -fsanitize=address
echo "Asan 1, strstr 1"
time ./slimcc ../c_files/gcc.c -E -o/dev/null
```

**Comment by ramosian-glider** [comment]
Does the behavior change if you run with and without ASAN_OPTIONS=intercept_strstr=0 ?

**Comment by fuhsnn** [comment]
> Does the behavior change if you run with and without ASAN_OPTIONS=intercept_strstr=0 ?

Yes, the slowdown is gone with `intercept_strstr=0`, thanks for the pointer!
So this issue is really about the intercept wrapper being slow for a simple short needle (`"*/"` in this case).

**Comment by ramosian-glider** [comment]
In addition to calling the original `strstr(haystack, needle)`, the interceptor does roughly the following:

```
  uptr len1 = internal_strlen(s1);
  uptr len2 = internal_strlen(s2);
  __asan_region_is_poisoned(s1, common_flags()->strict_string_checks ? (internal_strlen(s1)) + 1 : (r ? r - s1 + len2 : len1 + 1);)
  __asan_region_is_poisoned(s2, len2 + 1);
```

Maybe setting `strict_string_checks=0` should speed the things up?

@vitalybuka I think there's an opportunity here to spare a call to `internal_strlen(s1)`
